### PR TITLE
Docker image test sync

### PR DESCRIPTION
**Pull request by ScottBailey**
1. `bootstrap.sh` was updated to:
   - tag the generated image with the git short commit hash
   - optionally (`--release`) tag with release tags
     - optionally (`--push`) push to release location.
2. Moved version info into its own file (src/dunes/version.py).
3. Added a hidden `--short-version` flag to report JUST the version info (e.g. `1.1.0`).
4. A test was added to ensure either the short hash commit image is being used OR - if this was NOT built from a git repo - that the version image is being used.

### CODE REVIEW

File: bootstrap.sh [hunk 1]
```diff
@@ -3,21 +3,28 @@ set -o errexit -o pipefail -o noclobber -o nounset
 
 getopt --test && echo "'getopt --test' failed in this environment." && exit 1
 
-LONGOPTS=leap:,cdt:,help
-OPTIONS=l:c:h
+# Set SDIR to the path containing this script. Even if it's a symlink.
+SDIR=$(dirname $(readlink -f "$0"))
```
[the remark this answers is not in the document]
Comment (ScottBailey) [reply bootstrap.sh:7]: Added a `cd ${SDIR}` line to address this. ca4f015

File: bootstrap.sh [hunk 2]
```diff
@@ -49,3 +60,19 @@ if [[ $(uname) == "Darwin" ]]; then
 fi
 
 docker build --no-cache --build-arg USER_ID=0 --build-arg GROUP_ID="$GROUP_ID" $LEAP_ARGUMENT $CDT_ARGUMENT -f Dockerfile.unix -t dune "$SDIR"
+
+if [ -n "${RELEASE_VERSION}" ]; then
+    docker tag dune dune:${RELEASE_VERSION}
+    docker tag dune ghcr.io/antelopeio/dune:latest
+    docker tag dune ghcr.io/antelopeio/dune:${RELEASE_VERSION}
+    echo "Tagged image with latest and version '${RELEASE_VERSION}'."
```
[the remark this answers is not in the document]
Comment (ScottBailey) [reply bootstrap.sh:75]: Done, also added a `--push` arg to bootstrap.sh to make it even easier.

File: bootstrap.sh [hunk 3]
```diff
@@ -49,3 +60,19 @@ if [[ $(uname) == "Darwin" ]]; then
 fi
 
 docker build --no-cache --build-arg USER_ID=0 --build-arg GROUP_ID="$GROUP_ID" $LEAP_ARGUMENT $CDT_ARGUMENT -f Dockerfile.unix -t dune "$SDIR"
+
+if [ -n "${RELEASE_VERSION}" ]; then
+    docker tag dune dune:${RELEASE_VERSION}
+    docker tag dune ghcr.io/antelopeio/dune:latest
+    docker tag dune ghcr.io/antelopeio/dune:${RELEASE_VERSION}
```
[the remark this answers is not in the document]
Comment (ScottBailey) [reply bootstrap.sh]: dba6292 :-)

File: bootstrap.sh [hunk 4]
```diff
@@ -49,3 +60,19 @@ if [[ $(uname) == "Darwin" ]]; then
...
+    echo "Tagged image with latest and version '${RELEASE_VERSION}'."
+fi
+
+GIT_CMD="git -C ${SDIR} rev-parse --short HEAD"
+if ${GIT_CMD} > /dev/null 2> /dev/null; then
+    COMMIT_HASH=$(${GIT_CMD})
+    docker tag dune dune:${COMMIT_HASH}
```
[the remark this answers is not in the document]
Comment (ScottBailey) [reply bootstrap.sh]: Commit hash should NOT contain control or space characters, so this should be unnecessary. Please correct me if I'm wrong.

[the remark this answers is not in the document]
Comment (mikelik) [reply bootstrap.sh]: You are right.
However it doesn't do any harm to add it and `shellcheck` will not complain if someday we will add it to the CI ;)

File: bootstrap.sh [hunk 5]
```diff
@@ -3,21 +3,31 @@ set -o errexit -o pipefail -o noclobber -o nounset
 
 getopt --test && echo "'getopt --test' failed in this environment." && exit 1
 
-LONGOPTS=leap:,cdt:,help
-OPTIONS=l:c:h
+# Set SDIR to the path containing this script. Even if it's a symlink.
+SDIR=$(dirname $(readlink -f "$0"))
```
Comment (mikelik) [inline bootstrap.sh:7]:
```suggestion
SDIR=$(dirname "$(readlink -f "$0")")
```

File: tests/a010_ensure_python_version.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python3
```
Comment (mikelik) [inline tests/a010_ensure_python_version.py:1]: When I run `pytest tests` from root directory this test isn't executed. Do you have the same issue?
```
mikel@msi:~/repo/DUNE$ pytest tests/                                                                                                    
========================================================= test session starts ==========================================================
platform linux -- Python 3.8.10, pytest-7.2.1, pluggy-0.13.0                                                                            
rootdir: /home/mikel/repo/DUNE                                                                                                          
collected 31 items                                                                                                                     

tests/a100_ensure_container_image_test.py .                                                                                      [  3%]
tests/test_all_options_tested.py s                                                                                               [  6%]
tests/test_antler_proj.py s                                                                                                      [  9%]
tests/test_bootstrap.py .                                                                                                        [ 12%]
tests/test_bootstrap_full.py .                                                                                                   [ 16%]
tests/test_container.py .                                                                                                        [ 19%]
tests/test_debug.py .                                                                                                            [ 22%]
tests/test_deploy.py .                                                                                                           [ 25%]
tests/test_help.py ..                                                                                                            [ 32%]
tests/test_keys.py .                                                                                                             [ 35%]
tests/test_list_features.py .                                                                                                    [ 38%]
tests/test_monitor.py .                                                                                                          [ 41%]
tests/test_nodes.py ..                                                                                                           [ 48%]
tests/test_plugin.py ..                                                                                                          [ 54%]
tests/test_project.py ....                                                                                                       [ 67%]
tests/test_remainder.py ..                                                                                                       [ 74%]
tests/test_version.py .                                                                                                          [ 77%]
tests/test_version_selector.py .....                                                                                             [ 93%]
tests/test_wallet.py ..                                                                                                          [100%]

============================================== 29 passed, 2 skipped in 300.12s (0:05:00) ===============================================
```

File: bootstrap.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (ScottBailey) [reply bootstrap.sh]: Fixed in 4d45c36

File: bootstrap.sh (hunk 5 above)
Comment (ScottBailey) [reply bootstrap.sh]: This is a very good suggestion! Fixed in e1a49e9

File: tests/a010_ensure_python_version.py (hunk 6 above)
Comment (ScottBailey) [reply tests/a010_ensure_python_version.py]: Whoops! Fixed in 229ab2c.

The test must follow the format: `test_xxxx.py` or `xxxx_test.py`. At least for an automatic run, anyway.

**Review by mikelik — APPROVED** [review]
Good work, this saves someone's day of debugging :)

File: bootstrap.sh [hunk 7]
```diff
@@ -3,21 +3,31 @@ set -o errexit -o pipefail -o noclobber -o nounset
...
 PARSED=$(getopt --options=$OPTIONS --longoptions=$LONGOPTS --name "$0" -- "$@")
 
 eval set -- "$PARSED"
 LEAP_ARGUMENT=
 CDT_ARGUMENT=
+RELEASE_VERSION=
+PUSH_VERSION=
 
-usage="$(basename "$0") [-l|--leap=version] [-c|--cdt=version]
+usage="$(basename "$0") [-l|--leap=version] [-c|--cdt=version] [-r|--release]
```
Comment (pliuwu1998) [inline bootstrap.sh:21]: Isn't [-p|--push]missing in above usage= line?

**Review by jolly-fellow — APPROVED** [review]

**Review by pliuwu1998 — APPROVED** [review]

**Pull request merged**